spring-web/src/main/java/org/springframework/http/codec/json/AbstractJackson2Encoder.java

```diff
@@ -147,45 +147,56 @@ public Flux<DataBuffer> encode(Publisher<?> inputStream, DataBufferFactory buffe
 
 		if (inputStream instanceof Mono) {
 			return Mono.from(inputStream)
-					.map(value -> encodeValue(value, bufferFactory, elementType, mimeType, hints))
+					.flatMap(value -> createEncodingToolsForStream(value, elementType, mimeType, hints)
+							.map(tools -> encodeValue(value, tools.mapper(), tools.writer(),
+									bufferFactory, mimeType, hints)))
 					.flux();
 		}
 		else {
 			byte[] separator = getStreamingMediaTypeSeparator(mimeType);
 			if (separator != null) { // streaming
-				try {
-					ObjectMapper mapper = selectObjectMapper(elementType, mimeType);
-					if (mapper == null) {
-						throw new IllegalStateException("No ObjectMapper for " + elementType);
-					}
-					ObjectWriter writer = createObjectWriter(mapper, elementType, mimeType, null, hints);
-					ByteArrayBuilder byteBuilder = new ByteArrayBuilder(writer.getFactory()._getBufferRecycler());
-					JsonEncoding encoding = getJsonEncoding(mimeType);
-					JsonGenerator generator = mapper.getFactory().createGenerator(byteBuilder, encoding);
-					SequenceWriter sequenceWriter = writer.writeValues(generator);
-
-					return Flux.from(inputStream)
-							.map(value -> encodeStreamingValue(value, bufferFactory, hints, sequenceWriter, byteBuilder,
-									separator))
-							.doAfterTerminate(() -> {
-								try {
-									byteBuilder.release();
-									generator.close();
-								}
-								catch (IOException ex) {
-									logger.error("Could not close Encoder resources", ex);
-								}
-							});
-				}
-				catch (IOException ex) {
-					return Flux.error(ex);
+				ObjectMapper mapper = selectObjectMapper(elementType, mimeType);
+				if (mapper == null) {
+					return Flux.error(new IllegalStateException("No ObjectMapper for " + elementType));
```

Comment:
Another "breaking" change - maybe we should be OK with this given that it's a more appropriate handling IMO.

Reply:
Going to go with the "right way" and let the them tell me not to and I'll revert if they don't want.